### PR TITLE
Bug/828 getitem typechecking

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Description

Expanded type-checking of `key` within `getitem`, so that 1-element `array-like` objects are allowed. This includes 0-dim arrays.

Issue/s resolved: #828 

## Changes proposed:
- private function `__is_singular()` does the checking within `getitem`, thanks @ben-bou !
-
-
-

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by ClaudiaComito** [comment]
rerun tests

File: heat/core/dndarray.py [hunk 1]
```diff
@@ -730,6 +730,32 @@ def __getitem__(self, key: Union[int, Tuple[int, ...], List[int, ...]]) -> DNDar
             slices = [slice(None)] * (self.ndim - (len(kst) + len(kend)))
             key = kst + slices + kend
 
+        def __is_singular(obj: any) -> bool:
```
Comment (ben-bou) [inline heat/core/dndarray.py:733]: I'm sorry and don't want to cause more work, but honestly we already have the `self_proxy` at this point.
Something like the following could be 100 % torch-compatible (i've got no idea about performance or edge cases though)
```python
zeros = tuple([0] * (self.ndim -1))
is_singular = bool(1 - self_proxy[ *zeros[:i], key[i], *zeros[i:] ].ndim)
```

File: heat/core/tests/test_dndarray.py [hunk 2]
```diff
@@ -1069,7 +1069,7 @@ def test_setitem_getitem(self):
         # slice in 1st dim across 1 node (2nd) w/ singular second dim
         c = ht.zeros((13, 5), split=0)
         c[8:12, 1] = 1
-        b = c[8:12, 1]
+        b = c[8:12, np.array(1)]
```
Comment (ben-bou) [inline heat/core/tests/test_dndarray.py:1072]: Please also add one with `np.int64` instead of `np.array`

**Review by ben-bou — APPROVED** [review]

File: heat/core/dndarray.py [hunk 3]
```diff
@@ -730,6 +730,10 @@ def __getitem__(self, key: Union[int, Tuple[int, ...], List[int, ...]]) -> DNDar
             slices = [slice(None)] * (self.ndim - (len(kst) + len(kend)))
             key = kst + slices + kend
 
+        def __is_singular(key: any, axis: int, self_proxy: DNDarray = self_proxy) -> bool:
```
Comment (ben-bou) [inline heat/core/dndarray.py:733]: sanitize axis

**Review by ben-bou — APPROVED** [review]

**Comment by ben-bou** [comment]
A similar type-checking could be employed everywhere in the code where `isinstance(obj, int)` is used (e.g. [sanitize_axis](https://github.com/helmholtz-analytics/heat/blob/c62cd072479efac8d850b85fc9e426e75fafd344/heat/core/stride_tricks.py#L100))
There's no reason why a numpy integer should not be used to specify an axis.

File: heat/core/dndarray.py [hunk 4]
```diff
@@ -730,6 +730,10 @@ def __getitem__(self, key: Union[int, Tuple[int, ...], List[int, ...]]) -> DNDar
...
+            zeros = tuple([0] * (self.ndim - 1))
```
Comment (ben-bou) [inline heat/core/dndarray.py:734]: `self_proxy.ndim` is better because the way it is, `self_proxy` can be something else (even not actually a proxy of self)

File: heat/core/dndarray.py (hunk 3 above)
Comment (ben-bou) [reply heat/core/dndarray.py:733]: maybe rename `self_proxy` to `torch_proxy`

**Review by coquelin77 — CHANGES_REQUESTED** [review]

File: heat/core/dndarray.py [hunk 5]
```diff
@@ -730,6 +730,32 @@ def __getitem__(self, key: Union[int, Tuple[int, ...], List[int, ...]]) -> DNDar
...
+            except Exception:
+                pass
+            if hasattr(obj, "__len__") and len(obj) == 1:
+                return True
+            if hasattr(obj, "shape") and len(obj.shape) == 0:
+                return True
+            if hasattr(obj, "numel"):
+                return obj.numel() == 1
+            if hasattr(obj, "size"):
+                return obj.size == 1
```
[the remark this answers is not in the document]
Comment (ben-bou) [reply heat/core/dndarray.py]: probably not, but i wanted to be as duck-typing friendly as possible. anyway, utilizing torch is way better because supporting ducktypes and array-likes not supported by torch would only lead to new errors

File: heat/core/dndarray.py (hunk 3 above)
Comment (ben-bou) [reply heat/core/dndarray.py]: I think both depend on whether this code will ever be used non-internally. My reasoning is that, theoretically, `self_proxy` can be any torch.Tensor and the function will work, so why should it be named after `self`.

File: heat/core/dndarray.py [hunk 6]
```diff
@@ -887,6 +881,12 @@ def __getitem__(self, key: Union[int, Tuple[int, ...], List[int, ...]]) -> DNDar
             balanced=True if new_split is None else None,
         )
 
+    @staticmethod
+    def __getitem_is_key_singular(key: any, axis: int, self_proxy: torch.Tensor) -> bool:
```
Comment (ClaudiaComito) [inline heat/core/dndarray.py:885]: I agree with abstracting this method, I'm confused by the name. Something like `__key_is_singular` would work well I think, I'm sure we'll end up using this in `setitem` as well.

**Comment by coquelin77** [comment]
rerun tests

File: heat/core/dndarray.py (hunk 6 above)
Comment (coquelin77) [reply heat/core/dndarray.py:885]: okay that works for me. ill make the change now. my motivation behind the name was to keep it right below the getitem function

Comment (coquelin77) [reply heat/core/dndarray.py:885]: ive changed it to `__is_key_singular` to keep it nearby

**Review by coquelin77 — APPROVED** [review]
@ClaudiaComito I approve of your changes, but I think that you should also look at the minor changes that I made (i believe that it was primarily abstraction) before this is merged. I think that you should be the one to hit the button when you are ready to merge this

**Review by coquelin77 — APPROVED** [review]
